### PR TITLE
Provena PR (Minor): Model run edit - add study link

### PR DESCRIPTION
**Pull request by Rosspet**
# Provena PR (Minor): Model run edit - add study link

⚠️ DRAFT ⚠️ because there is still UI work to complete this feature

## Checklist

-   [ ] If tests are required for this change, are they implemented?
-   [ ] Are user documentation changes required, if so, is there a task to track it and/or is it completed?
-   [ ] If migrations are required, is the process documented below?
-   [ ] If developer/system documentation updates are required, is there a task to track it and/or is it completed?
-   [ ] At least one developer has reviewed this change (unless PR is being used to mark a commit point without need for review)?
-   [ ] If change requires update to the [manual testing runsheet](https://confluence.csiro.au/pages/viewpage.action?pageId=1701138616), is there a task to track it and/or is it completed?

## Description

This adds a new endpoint into the provenance api for editing an existing model run record to add a link to an existing study. 

This was identified as a feature request from users who enjoy organising (sometimes already registered) model run efforts into "studies". 

The endpoint first updates the model run record entity in the registry to include the study link in the record info, then upon success, will update the provenance graph using the job system LODGE MODEL RUN ONLY job type. This job type does not create new entities, just a new record in the graph in the form of an overwrite, so new graph nodes are not added either, just the missing edge between the study and the model run (though if the study doesn't already exist somewhere in the graph then it'll be added of course).

## Notes for reviewer

Up to reviewer if tests are required for the model run metadata edit and/or prov graph edit.

### CODE REVIEW

File: prov-api/routes/model_run/model_run.py [hunk 1]
```diff
@@ -2,17 +2,21 @@
...
 from helpers.auth_helpers import *
-from helpers.job_api_helpers import submit_model_run_lodge_job, submit_batch_lodge_job
-from ProvenaInterfaces.AsyncJobModels import ProvLodgeModelRunPayload, ProvLodgeBatchSubmitPayload
+from helpers.job_api_helpers import submit_model_run_lodge_job, submit_batch_lodge_job, submit_model_run_lodge_only_job
+from ProvenaInterfaces.AsyncJobModels import ProvLodgeModelRunPayload, ProvLodgeBatchSubmitPayload, ProvLodgeModelRunLodgeOnlyPayload
 from config import get_settings, Config
 
 router = APIRouter()
 
+edit_root_prefix = "/edit"
```
Comment (PeterBaker0) [inline prov-api/routes/model_run/model_run.py:18]: can you make this CAPS_CASE to indicate it's a constant

File: prov-api/routes/model_run/model_run.py [hunk 2]
```diff
@@ -164,3 +168,71 @@ async def register_model_run_sync(
...
+    request_style = RequestStyle(
+        user_direct=roles.user, service_account=None)
+    
+    model_run, study = await validate_model_run_study_linking(
+        model_run_id=model_run_id,
+        study_id=study_id,
+        request_style=request_style,
+        config=config
+    )
+    model_run: ItemModelRun
```
Comment (PeterBaker0) [inline prov-api/routes/model_run/model_run.py:191]: what is this doing?

File: prov-api/routes/model_run/model_run.py [hunk 3]
```diff
@@ -164,3 +168,71 @@ async def register_model_run_sync(
...
+        ) from e
+
+    # done
+    return AddStudyLinkResponse(
+        status=Status(
+            success=True, details=f"Succesfully linked model run and study together in the registry and lodged a job for the provenance graph update with given session ID."),
+        model_run_id=model_run_id,
+        study_id=study_id,
+        session_id=session_id
+    )
```
Comment (PeterBaker0) [inline prov-api/routes/model_run/model_run.py:238]: nice

**Review by PeterBaker0 — COMMENTED** [review]
Looks great - have you tested the functionality manually?

File: prov-api/routes/model_run/model_run.py (hunk 2 above)
Comment (Rosspet) [reply prov-api/routes/model_run/model_run.py:191]: image
for some reason my mypy typing wasnt working properly so just added that simple line to tell it what it is i think its allg

File: prov-api/routes/model_run/model_run.py (hunk 1 above)
Comment (Rosspet) [reply prov-api/routes/model_run/model_run.py:18]: yes - good point

**Comment by Rosspet** [comment]
> Looks great - have you tested the functionality manually?

Yes I've tested it manually - also have tested the incorrect input behaviour. For example, trying to link a model run that is already linked, or supplying IDs for items that dont exist or are not the correct item types. Correct behaviour is observed.

**Review by jyucsiro — APPROVED** [review]
Tested the feature and confirmed it works as designed and expected.

**Pull request merged**